planex/cmd/fetch.py

```diff
@@ -115,35 +103,42 @@ def fetch_http(url, filename, retries):
     Download the file at url and store it as filename
     """
 
-    while True:
-        retries -= 1
-        try:
-            url_string = urlparse.urlunparse(url)
-            logging.debug("Fetching %s to %s", url_string, filename)
+    url_string = urlparse.urlunparse(url)
+    logging.debug("Fetching %s to %s", url_string, filename)
+
+    useragent = "planex-fetch/%s" % pkg_resources.require("planex")[0].version
+
+    # We need this because centos ships requests 2.8.0
+    # It can be greatly simplified with requests >= 2.18.0
+    headers = requests.utils.default_headers()
+    headers.update({
+        "user-agent": useragent,
+    })
+
+    # Once we use requests >= 2.18.0, we should change this into
+    # with requests.get ... as r:
+    req = requests_retry_session(retries).get(
+        url_string, headers=headers, timeout=30, stream=True)
+    req.raise_for_status()
 
-            tmp_filename = filename + "~"
-            with open(tmp_filename, "wb") as tmp_file:
-                curl_get(url_string, tmp_file)
-                best_effort_file_verify(tmp_filename)
-                shutil.move(tmp_filename, filename)
-                # Write an origin file for tracking.
-                with open('{0}.origin'.format(filename), 'w') as origin_file:
-                    origin_file.write('{0}\n'.format(url_string))
-                return
+    tmp_filename = filename + "~"
+    with open(tmp_filename, 'wb') as out:
+        shutil.copyfileobj(req.raw, out)
+    best_effort_file_verify(tmp_filename)
+    shutil.move(tmp_filename, filename)
```

Comment:
It would be great to get rid of this as it's racy: #208.   We don't need to do anything special with the output file in fetch to clean up after a failed download because we define .DELETE_ON_ERROR in Makefile.rules, so if fetch returns a nonzero exit code make will delete the output file anyway.

Reply:
I am all for it, but if we write into filename directly this is still racy as both instances will try to write over the same file

Reply:
How can we prevent `planex-fetch` to fetch the same file multiple times concurrently?